File: backend/internal/dto/lote.go
package dto

import (
	"time"

	"github.com/google/uuid"
)

// CreateLoteRequest representa o payload de criacao de um novo lote.
type CreateLoteRequest struct {
	GalpaoID          uuid.UUID `json:"galpao_id" validate:"required"`
	DataAlojamento    string    `json:"data_alojamento" validate:"required"` // formato: YYYY-MM-DD
	DataPrevistaAbate string    `json:"data_prevista_abate" validate:"required"` // formato: YYYY-MM-DD
	Quantidade        int       `json:"quantidade" validate:"required,gt=0"`
	Tipo              string    `json:"tipo" validate:"required,oneof=Femea Macho Misto"`
	Linhagem          *string   `json:"linhagem,omitempty"`
	PesoInicialG      *float64  `json:"peso_inicial_g,omitempty" validate:"omitempty,gt=0"`
}

// LoteResponse representa a resposta de um lote com dados basicos.
type LoteResponse struct {
	ID                uuid.UUID  `json:"id"`
	UsuarioID         uuid.UUID  `json:"usuario_id"`
	GalpaoID          uuid.UUID  `json:"galpao_id"`
	GalpaoNome        string     `json:"galpao_nome"`
	DataAlojamento    time.Time  `json:"data_alojamento"`
	DataPrevistaAbate time.Time  `json:"data_prevista_abate"`
	Quantidade        int        `json:"quantidade"`
	Tipo              string     `json:"tipo"`
	Linhagem          *string    `json:"linhagem,omitempty"`
	PesoInicialG      float64    `json:"peso_inicial_g"`
	Status            string     `json:"status"`
	DiasDeVida        int        `json:"dias_de_vida"`
	AvesVivas         int        `json:"aves_vivas"`
	UltimoPesoMedio   *float64   `json:"ultimo_peso_medio,omitempty"`
	MortalidadeTotal  int        `json:"mortalidade_total"`
	DataFinalizacao   *time.Time `json:"data_finalizacao,omitempty"`
	CreatedAt         time.Time  `json:"created_at"`
	UpdatedAt         time.Time  `json:"updated_at"`
}

// LoteIndicadoresResponse representa os indicadores zootecnicos do lote.
type LoteIndicadoresResponse struct {
	LoteID           uuid.UUID  `json:"lote_id"`
	DiasDeVida       int        `json:"dias_de_vida"`
	AvesVivas        int        `json:"aves_vivas"`
	AvesAlojadas     int        `json:"aves_alojadas"`
	MortalidadeTotal int        `json:"mortalidade_total"`
	MortalidadePct   float64    `json:"mortalidade_pct"`
	Viabilidade      float64    `json:"viabilidade"`

	// Peso
	UltimoPesoMedio  *float64 `json:"ultimo_peso_medio,omitempty"`
	PesoPadraoG      *float64 `json:"peso_padrao_g,omitempty"`
	DesvioPesoPct    *float64 `json:"desvio_peso_pct,omitempty"`

	// Indices zootecnicos
	GPD              *float64 `json:"gpd,omitempty"`              // Ganho de peso diario (g/dia)
	ICA              *float64 `json:"ica,omitempty"`              // Indice de conversao alimentar
	IEA              *float64 `json:"iea,omitempty"`              // Indice de eficiencia alimentar
	IEP              *float64 `json:"iep,omitempty"`              // Indice de eficiencia produtiva
	ClassificacaoIEP *string  `json:"classificacao_iep,omitempty"` // Ruim, Regular, Bom, Excelente

	// Racao
	ConsumoTotalRacaoKg  *float64 `json:"consumo_total_racao_kg,omitempty"`
	RecebidoTotalRacaoKg *float64 `json:"recebido_total_racao_kg,omitempty"`
	SaldoRacaoKg         *float64 `json:"saldo_racao_kg,omitempty"`
	DiasEstoqueRacao     *float64 `json:"dias_estoque_racao,omitempty"`
	ConsumoRacaoPorAve   *float64 `json:"consumo_racao_por_ave,omitempty"` // kg/ave

	// Agua
	ConsumoAguaDiaL      *float64 `json:"consumo_agua_dia_l,omitempty"`
	RelacaoAguaRacao     *float64 `json:"relacao_agua_racao,omitempty"`

	// Projecao
	ProjecaoPeso         *ProjecaoPesoDTO `json:"projecao_peso,omitempty"`
}

// ProjecaoPesoDTO representa a projecao de peso do lote.
type ProjecaoPesoDTO struct {
	PesoEstimadoG float64   `json:"peso_estimado_g"`
	DataEstimada  time.Time `json:"data_estimada"`
	GPDG          float64   `json:"gpd_g"`
	Mensagem      string    `json:"mensagem"`
}

// FinalizarLoteRequest representa o payload para finalizar um lote.
type FinalizarLoteRequest struct {
	DataFinalizacao   string   `json:"data_finalizacao" validate:"required"` // formato: YYYY-MM-DD
	AvesEntregues     int      `json:"aves_entregues" validate:"required,gt=0"`
	PesoTotalEntregue float64  `json:"peso_total_entregue" validate:"required,gt=0"`
	SobraRacaoKg      *float64 `json:"sobra_racao_kg,omitempty" validate:"omitempty,gte=0"`
}

File: flutter_app/lib/features/lote_detail/domain/entities/lote.dart
/// Entidade Lote para o feature lote_detail.
///
/// Representa um lote de aves com seus dados basicos, mapeando
/// exatamente os campos retornados pelo backend Go (LoteResponse DTO).
class Lote {
  final String id;
  final String? usuarioId;
  final String galpaoId;
  final String? galpaoNome;
  final String dataAlojamento;
  final String? dataPrevistaAbate;
  final int quantidade;
  final String tipo; // 'Femea', 'Macho', 'Misto'
  final String? linhagem;
  final double pesoInicialG;
  final String status; // 'ativo', 'finalizado'
  final int? diasDeVida;
  final int? avesVivas;
  final double? ultimoPesoMedio;
  final int? mortalidadeTotal;
  final String? dataFinalizacao;
  final String createdAt;
  final String updatedAt;

  const Lote({
    required this.id,
    this.usuarioId,
    required this.galpaoId,
    this.galpaoNome,
    required this.dataAlojamento,
    this.dataPrevistaAbate,
    required this.quantidade,
    required this.tipo,
    this.linhagem,
    required this.pesoInicialG,
    required this.status,
    this.diasDeVida,
    this.avesVivas,
    this.ultimoPesoMedio,
    this.mortalidadeTotal,
    this.dataFinalizacao,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Lote.fromJson(Map<String, dynamic> json) {
    return Lote(
      id: json['id'] as String,
      usuarioId: json['usuario_id'] as String?,
      galpaoId: json['galpao_id'] as String,
      galpaoNome: json['galpao_nome'] as String?,
      dataAlojamento: json['data_alojamento'] as String,
      dataPrevistaAbate: json['data_prevista_abate'] as String?,
      quantidade: (json['quantidade'] as num).toInt(),
      tipo: json['tipo'] as String? ?? 'Misto',
      linhagem: json['linhagem'] as String?,
      pesoInicialG: (json['peso_inicial_g'] as num?)?.toDouble() ?? 0.0,
      status: json['status'] as String,
      diasDeVida: (json['dias_de_vida'] as num?)?.toInt(),
      avesVivas: (json['aves_vivas'] as num?)?.toInt(),
      ultimoPesoMedio: (json['ultimo_peso_medio'] as num?)?.toDouble(),
      mortalidadeTotal: (json['mortalidade_total'] as num?)?.toInt(),
      dataFinalizacao: json['data_finalizacao'] as String?,
      createdAt: json['created_at'] as String,
      updatedAt: json['updated_at'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      if (usuarioId != null) 'usuario_id': usuarioId,
      'galpao_id': galpaoId,
      if (galpaoNome != null) 'galpao_nome': galpaoNome,
      'data_alojamento': dataAlojamento,
      if (dataPrevistaAbate != null) 'data_prevista_abate': dataPrevistaAbate,
      'quantidade': quantidade,
      'tipo': tipo,
      if (linhagem != null) 'linhagem': linhagem,
      'peso_inicial_g': pesoInicialG,
      'status': status,
      if (diasDeVida != null) 'dias_de_vida': diasDeVida,
      if (avesVivas != null) 'aves_vivas': avesVivas,
      if (ultimoPesoMedio != null) 'ultimo_peso_medio': ultimoPesoMedio,
      if (mortalidadeTotal != null) 'mortalidade_total': mortalidadeTotal,
      if (dataFinalizacao != null) 'data_finalizacao': dataFinalizacao,
      'created_at': createdAt,
      'updated_at': updatedAt,
    };
  }
}
